src/packages/permissions/permissions.py

```diff
@@ -16,18 +16,51 @@
 from functools import wraps
 from typing import Any, Union, Callable, Dict, Set
 
-from config import config
+from src.config import CONFIG_MARKER
 from ..context import Context
 
 LOG = logging.getLogger(f"mecha.{__name__}")
 
 
+@CONFIG_MARKER
+def validate_config(data: Dict):
+    """
+    Validate new configuration data.
+
+    Args:
+        data (typing.Dict): new configuration data  to validate
+
+    Raises:
+        ValueError:  config section failed to validate.
+        KeyError:  config section failed to validate.
+    """
+    ...  # TODO: permissions config validation
+
+
+@CONFIG_MARKER
+def rehash_handler(data: Dict):
+    """
+    Apply new configuration data
+
+    Args:
+        data (typing.Dict): new configuration data to apply.
+
+    """
+    LOG.debug("applying new permissions scheme...")
+    RECRUIT.update(data['permissions']['recruit'])
+    RAT.update(data['permissions']['rat'])
+    OVERSEER.update(data['permissions']['overseer'])
+    TECHRAT.update(data['permissions']['techrat'])
+    ADMIN.update(data['permissions']['administrator'])
+
+
 class Permission:
     """
     A permission level
     """
+    __slots__ = ["level", "_vhosts", "_denied_message"]
 
-    def __init__(self, level: int, vhosts: Set[str],
+    def __init__(self, level: int = -1, vhosts: Set[str] = None,
```

Comment:
[Style] I've always had it drilled into my head that magic numbers are bad. If `-1` is meant to mean something special, it should be assigned to a constant variable that can then be documented.

Reply:
Its not really a magic number, and i feel it requires no documentation.

Without specific initialization, all permission objects have a default permission level of -1, which is lower than any assigned value. 

I suppose this is a "fail safe" default, rather than a "fail secure".
Optimally, this value is immediately overwritten during post-init & during subsequent configuration file reloads.

My question here is should we "fail secure" and set this to be a very large integer value rather than a out-of-range small integer.